modules/8_reinforcement_learning/reinforcement_learning.py

```diff
@@ -0,0 +1,144 @@
+        print('best_action is not implemented')
+        exit(1)
+
+    def update_q_values(self, observation, action, reward, next_observation, done):
+        """Update stored q_values with provided data following Bellman equation.
+
+        The updated value is a weighted sum of the old value taken with 1.0 - ALPHA coefficient and the new one with
+        ALPHA. The new value is a sum of reward recieved and what the agent thinks about the next obesrvation discounted
+        by GAMMA like if it played the best action.
+        
+        If done==True the update doesn't care about next_observation.
+        """
+        print('update_q_values is not implemented')
+        exit(1)
+
+
+def check_agent_actions(agent):
+    assert np.issubdtype(agent.best_action(0).dtype, np.integer)
+    assert np.issubdtype(agent.best_action(15).dtype, np.integer)
+    assert 0 <= agent.best_action(1) <= 3
+
+
+def check_agent_update():
+    agent = Agent(16, 4)
+    assert (agent.q_values == 0).all()
+    agent.update_q_values(0, 0, 9e9, 0, True)
+    assert agent.q_values[0, 0] > 0
+
+
+def final_test(agent):
+    EPISODE_NUM = 1000
+    env = wrappers.TimeLimit(toy_text.FrozenLakeEnv(map_name='8x8'), max_episode_steps=1000)
+    score_sum = 0.0
+    for _ in range(EPISODE_NUM):
+        observation = env.reset()
+        while True:
+            observation, reward, done, _ = env.step(agent.best_action(observation))
+            score_sum += reward
+            if done: break
+    assert score_sum / EPISODE_NUM >= 0.99
+
+
+def train():
+    UPDATE_PERIOD = 10000
+    EPSILON_DECAY = 0.7
+    env = wrappers.TimeLimit(toy_text.FrozenLakeEnv(map_name='8x8'), max_episode_steps=1000)
+
+    print('agent is not created')
+    exit(1)
+    agent = None
+
+    check_agent_actions(agent)
+    check_agent_update()
+
+    epsilon = 1.0
+    last_scores_sum = 0.0
+    for episode_id in itertools.count():
+        score = 0.0
+        observation = env.reset()
+        while True:
+
+            print('perform a random action with probability of epsilon. Request best action from agent otherwise')
+            exit(1)
+            action = None
+
+            next_observation, reward, done, _ = env.step(action)
+            agent.update_q_values(observation, action, reward, next_observation, done)
+            score += reward
+            if done:
+                break
+            observation = next_observation
+
+        last_scores_sum += score
+        if episode_id > 0 and episode_id % UPDATE_PERIOD == 0:
+            epsilon *= EPSILON_DECAY
+            last_mean_score = last_scores_sum / UPDATE_PERIOD
+            print('last_mean_score:', last_mean_score)
+            if last_mean_score > 0.999:
+                print('You won!')
+                break
+            else:
+                last_scores_sum = 0.0
+    return agent
+
+
+def main():
+    try_gym()
+    agent = train()
+    final_test(agent)
```

Comment:
Do we need a test here if we have a separate file?

Reply:
This makes sure the agent will work even if test was not run locally. But I can remove it. Should I?

Reply:
I think it's better, yes. So `def final_test(agent)` will also be defined in the `test_reinforcement_learning`

Reply:
Updated